code/datums/components/trader/trader.dm

```diff
@@ -0,0 +1,413 @@
+	if(!pick || !can_trade(customer))
+		return
+	var/obj/item/item_to_buy = display_names[pick]
+	trader.face_atom(customer)
+	product_info = products[item_to_buy]
+	if(!product_info[TRADER_PRODUCT_INFO_QUANTITY])
+		trader.say("[initial(item_to_buy.name)] appears to be out of stock.")
+		return
+	trader.say("It will cost you [product_info[TRADER_PRODUCT_INFO_PRICE]] [trader_data.currency_name] to buy \the [initial(item_to_buy.name)]. Are you sure you want to buy it?")
+	var/list/npc_options = list(
+		"Yes" = radial_icons_cache[TRADER_RADIAL_YES],
+		"No" = radial_icons_cache[TRADER_RADIAL_NO],
+	)
+	var/buyer_will_buy = show_radial_menu(customer, trader, npc_options, custom_check = CALLBACK(src, PROC_REF(check_menu), customer), require_near = TRUE, tooltips = TRUE)
+	if(buyer_will_buy != "Yes" || !can_trade(customer))
+		return
+	trader.face_atom(customer)
+	if(!spend_buyer_offhand_money(customer, product_info[TRADER_PRODUCT_INFO_PRICE]))
+		trader.say(trader_data.return_trader_phrase(NO_CASH_PHRASE))
+		return
+	item_to_buy = new item_to_buy(get_turf(customer))
+	customer.put_in_hands(item_to_buy)
+	playsound(trader, trader_data.sell_sound, 50, TRUE)
+	log_econ("[item_to_buy] has been sold to [customer] (typepath used for product info; [item_to_buy.type]) by [trader] for [product_info[TRADER_PRODUCT_INFO_PRICE]] cash.")
+	product_info[TRADER_PRODUCT_INFO_QUANTITY] -= 1
+	trader.say(trader_data.return_trader_phrase(BUY_PHRASE))
+
+///Calculates the value of money in the hand of the buyer and spends it if it's sufficient
+/datum/component/trader/proc/spend_buyer_offhand_money(mob/customer, the_cost)
+	var/value = 0
+	var/obj/item/holochip/cash = customer.is_holding_item_of_type(/obj/item/holochip)
+	if(cash)
+		value += cash.credits
+	if((value >= the_cost) && cash)
+		return cash.spend(the_cost)
+	return FALSE //Purchase unsuccessful
+
+/**
+ * Tries to call sell_item on one of the customer's held items, if fail gives a chat message
+ *
+ * Gets both items in the customer's hands, and then tries to call sell_item on them, if both fail, he gives a chat message
+ * Arguments:
+ * * customer - (Mob REF) The mob trying to sell something
+ */
+/datum/component/trader/proc/try_sell(mob/customer)
+	if(!can_trade(customer))
+		return
+	var/mob/living/trader = parent
+	var/sold_item = FALSE
+	for(var/obj/item/an_item in customer.held_items)
+		if(sell_item(customer, an_item))
+			sold_item = TRUE
+			break
+	if(!sold_item && !can_trade(customer)) //don't talk if you are dead or fighting
+		trader.say(trader_data.return_trader_phrase(ITEM_REJECTED_PHRASE))
+
+
+/**
+ * Checks if an item is in the list of wanted items and if it is after a Yes/No radial returns generate_cash with the value of the item for the NPC
+ * Arguments:
+ * * customer - (Mob REF) The mob trying to sell something
+ * * selling - (Item REF) The item being sold
+ */
+/datum/component/trader/proc/sell_item(mob/customer, obj/item/selling)
+	var/mob/living/trader = parent
+	var/cost
```

Comment:
Same here, allocate these after we're done deciding if we are going to early return
`cost` is defined _way_ above where it actually gets any values put into it (why even predefine it as null before that point actually?)